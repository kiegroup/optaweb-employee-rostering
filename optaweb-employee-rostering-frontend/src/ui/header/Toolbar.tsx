/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Button,
  ButtonVariant,
  Dropdown,
  DropdownItem,
  DropdownToggle,
  Toolbar,
  ToolbarGroup,
  ToolbarItem,
} from '@patternfly/react-core';
import { BellIcon, CogIcon } from '@patternfly/react-icons';
import { Tenant } from 'domain/Tenant';
import * as React from 'react';
import { connect } from 'react-redux';
import { tenantOperations } from 'store/tenant';
import { AppState } from 'store/types';
import { withRouter, RouteComponentProps } from 'react-router-dom';

interface StateProps {
  currentTenantId: number;
  tenantList: Tenant[];
}

const mapStateToProps = ({ tenantData }: AppState): StateProps => ({
  currentTenantId: tenantData.currentTenantId,
  tenantList: tenantData.tenantList,
});

interface ToolbarState {
  isTenantSelectOpen: boolean;
}

export interface DispatchProps {
  refreshTenantList: typeof tenantOperations.refreshTenantList;
  changeTenant: typeof tenantOperations.changeTenant;
}

const mapDispatchToProps: DispatchProps = {
  refreshTenantList: tenantOperations.refreshTenantList,
  changeTenant: tenantOperations.changeTenant,
};

export type Props = RouteComponentProps & StateProps & DispatchProps;

export class ToolbarComponent extends React.Component<Props, ToolbarState> {
  constructor(props: Props) {
    super(props);
    this.state = { isTenantSelectOpen: false };
  }

  componentDidMount() {
    this.props.refreshTenantList();
  }

  setCurrentTenant(newTenantId: number) {
    this.setState({
      isTenantSelectOpen: false,
    });
    this.props.changeTenant({ tenantId: newTenantId, routeProps: this.props });
  }

  setIsTenantSelectOpen(isOpen: boolean) {
    this.setState({ isTenantSelectOpen: isOpen });
  }

  render() {
    const bellAndCog = (
      <ToolbarGroup>
        <ToolbarItem>
          <Button
            aria-label="Notifications"
            variant={ButtonVariant.plain}
          >
            <BellIcon />
          </Button>
        </ToolbarItem>
        <ToolbarItem>
          <Button
            aria-label="Settings"
            variant={ButtonVariant.plain}
            data-cy="settings"
            onClick={() => this.props.history.push('/admin')}
          >
            <CogIcon />
          </Button>
        </ToolbarItem>
      </ToolbarGroup>
    );
    const { tenantList, currentTenantId } = this.props;
    const { isTenantSelectOpen } = this.state;
    if (tenantList.length === 0) {
      return (
        <Toolbar>
          <ToolbarGroup />
          {bellAndCog}
        </Toolbar>
      );
    }

    const currentTenant = tenantList.find(t => t.id === currentTenantId) as Tenant;
    return (
      <Toolbar>
        <ToolbarGroup>
          <ToolbarItem>
            <Dropdown
              isPlain
              position="right"
              onSelect={event => event && this.setCurrentTenant(
                parseInt((event.target as HTMLElement).dataset.tenantid as string, 10),
              )}
              isOpen={isTenantSelectOpen}
              toggle={(
                <DropdownToggle onToggle={() => this.setIsTenantSelectOpen(!isTenantSelectOpen)}>
                  {currentTenant.name}
                </DropdownToggle>
              )}
              dropdownItems={tenantList.map(tenant => (
                <DropdownItem data-tenantid={tenant.id} key={tenant.id}>
                  {tenant.name}
                </DropdownItem>
              ))}
            />
          </ToolbarItem>
        </ToolbarGroup>
        {bellAndCog}
      </Toolbar>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(ToolbarComponent));
