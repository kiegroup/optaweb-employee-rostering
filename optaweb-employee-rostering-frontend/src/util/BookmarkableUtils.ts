
import { RouteComponentProps } from 'react-router';

export type UrlProps<T extends string> = { [K in T]: string|null };

const pathnameToQueryStringMap: Map<string, string> = new Map();

export function setTenantIdInUrl(props: RouteComponentProps, tenantId: number) {
  const endOfTenantId = props.location.pathname.indexOf('/', 1);
  if (endOfTenantId !== -1) {
    props.history.push(`/${tenantId}${props.location.pathname
      .slice(endOfTenantId)}`);
  }
  // Else, the page is not specific to a tenant, so we do nothing
}

export function getPropsFromUrl<T extends UrlProps<string> >(props: RouteComponentProps, defaultValues: T): T {
  const out: { [index: string]: string|null } = { ...defaultValues };
  if (props.location.search === '' && pathnameToQueryStringMap.has(props.location.pathname)) {
    const searchParams = new URLSearchParams(pathnameToQueryStringMap.get(props.location.pathname));
    // eslint-disable-next-line no-return-assign
    searchParams.forEach((value, key) => out[key] = value);
    // defer updating URL since that counts as an update operation and
    // you cannot update in render
    setTimeout(() => requestAnimationFrame(() => setPropsInUrl(props, out as T)));
  } else {
    const searchParams = new URLSearchParams(props.location.search);
    // eslint-disable-next-line no-return-assign
    searchParams.forEach((value, key) => out[key] = value);
  }
  return out as T;
}

export function setPropsInUrl<T extends UrlProps<string> >(props: RouteComponentProps, urlProps: Partial<T>) {
  const searchParams = new URLSearchParams(props.location.search);

  Object.keys(urlProps).forEach((key) => {
    const value = urlProps[key] as string|null|undefined;
    if (value !== undefined) {
      if (value !== null && value.length > 0) {
        searchParams.set(key, value);
      } else {
        searchParams.delete(key);
      }
    }
  });
  pathnameToQueryStringMap.set(props.location.pathname, `?${searchParams.toString()}`);
  props.history.push(`${props.location.pathname}?${searchParams.toString()}`);
}
