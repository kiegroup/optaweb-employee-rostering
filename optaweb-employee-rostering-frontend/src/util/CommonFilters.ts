
import { Predicate } from 'types';

export function stringFilter<T>(...mappers: ((obj: T) => string|string[])[]): (filter: string) => Predicate<T> {
  return filter => obj => mappers.find((mapper) => {
    const value = mapper(obj);
    return (typeof value === 'string') ? value.toLowerCase().includes(filter.toLowerCase())
      : value.find(v => v.toLowerCase().includes(filter.toLowerCase())) !== undefined;
  }) !== undefined;
}
