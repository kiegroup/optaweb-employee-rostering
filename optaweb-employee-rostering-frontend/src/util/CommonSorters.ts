
import { Sorter } from 'types';

export function stringSorter<T>(mapToString: (obj: T) => string): Sorter<T> {
  // eslint-disable-next-line no-nested-ternary
  return (a, b) => ((mapToString(a).toLowerCase() < mapToString(b).toLowerCase()) ? -1
    : (mapToString(a).toLowerCase() > mapToString(b).toLowerCase()) ? 1 : 0);
}
