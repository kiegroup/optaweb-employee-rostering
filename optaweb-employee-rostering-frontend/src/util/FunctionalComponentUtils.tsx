import { List } from 'immutable';
import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router';
import { Sorter } from 'types';
import { DataTableUrlProps } from 'ui/components/DataTable';

export function useUrlState(urlProperty: string, initialValue?: string):
[string|null, (newValue: string|null) => void] {
  const history = useHistory();
  const searchParams = new URLSearchParams(history.location.search);
  const [state, setState] = useState(searchParams.get(urlProperty) || initialValue || null);
  return [
    state || null,
    (newValue) => {
      setState(newValue);
      const newSearchParams = new URLSearchParams(history.location.search);
      if (newValue !== null) {
        newSearchParams.set(urlProperty, newValue);
      } else {
        newSearchParams.delete(urlProperty);
      }
      history.push(`${history.location.pathname}?${newSearchParams.toString()}`);
    }];
}

// From https://overreacted.io/making-setinterval-declarative-with-react-hooks/
export function useInterval(callback: Function, delay: number|null) {
  const savedCallback = useRef<Function>();

  // Remember the latest callback.
  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  // Set up the interval.
  // eslint-disable-next-line consistent-return
  useEffect(() => {
    function tick() {
      (savedCallback.current as Function)();
    }
    if (delay !== null) {
      const id = setInterval(tick, delay);
      return () => clearInterval(id);
    }
  }, [delay]);
}


export interface PaginationData<T> {
  filterText: string;
  page: number;
  itemsPerPage: number;
  filteredRows: List<T>;
  numOfFilteredRows: number;
  rowsInPage: List<T>;
  isReversed: boolean;
}

export function usePageableData<T>(urlProps: Omit<DataTableUrlProps, 'sortBy'>, data: T[],
  valueToText: (value: T) => string[], sortBy: Sorter<T>): PaginationData<T> {
  const filterText = urlProps.filter || '';
  const page = parseInt(urlProps.page || '1', 10);
  const itemsPerPage = parseInt(urlProps.itemsPerPage || '10', 10);
  const isReversed = urlProps.asc === 'false';

  const sortedRows = List.of(...data)
    .sort(isReversed ? (a, b) => sortBy(b, a) : sortBy);

  const filteredRows = sortedRows
    .filter(value => valueToText(value)
      .filter(text => text.toLowerCase().indexOf(filterText.toLowerCase()) !== -1).length > 0);

  const numOfFilteredRows = filteredRows.size;

  const rowsInPage = filteredRows
    .skip((page - 1) * itemsPerPage)
    .take(itemsPerPage);

  return {
    filterText,
    page,
    itemsPerPage,
    filteredRows,
    numOfFilteredRows,
    rowsInPage,
    isReversed,
  };
}
