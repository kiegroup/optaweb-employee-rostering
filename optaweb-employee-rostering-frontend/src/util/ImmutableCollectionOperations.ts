
import { DomainObject } from 'domain/DomainObject';
import DomainObjectView from 'domain/DomainObjectView';
import { Map, Seq } from 'immutable';

export function toggleElement<T>(collection: T[], element: T): T[] {
  if (collection.indexOf(element) !== -1) {
    return collection.filter(e => e !== element);
  }

  return [...collection, element];
}

interface ObjectWithKeys {
  [key: string]: any;
}

export function objectWithout<F>(obj: F, ...without: (keyof F)[]): F {
  const copy: ObjectWithKeys = {
    ...obj,
  };
  without.forEach(key => delete copy[key as string]);
  return copy as F;
}

export function mapDomainObjectToView<T extends DomainObject>(obj: T|DomainObjectView<T>): DomainObjectView<T> {
  const result = {} as ObjectWithKeys;
  const objWithKeys = obj as ObjectWithKeys;
  Object.keys(objWithKeys).forEach((key) => {
    if (objWithKeys[key] !== null && objWithKeys[key].id !== undefined) {
      result[key] = objWithKeys[key].id;
    } else if (Array.isArray(objWithKeys[key]) && objWithKeys[key].length > 0
        && objWithKeys[key].find((item: any) => item && item.id !== undefined) !== undefined) {
      result[key] = objWithKeys[key].map((ele: DomainObject) => ele.id);
    } else {
      result[key] = objWithKeys[key];
    }
  });
  return result as DomainObjectView<T>;
}

export function createIdMapFromList<T extends DomainObject>(collection: T[]): Map<number, DomainObjectView<T>> {
  let map = Map<number, DomainObjectView<T>>();
  collection.forEach((ele) => { (map = map.set(ele.id as number, mapDomainObjectToView(ele))); });
  return map;
}

export interface FluentValue<V, M> {
  result: V;
  then: (map: M) => FluentValue<V, M>;
}
export function conditionally<K, V>(seq: Seq<K, V>, mapper: (seq: Seq<K, V>) => Seq<K, V> | undefined):
FluentValue<Seq<K, V>, (seq: Seq<K, V>) => Seq<K, V> | undefined> {
  const out = mapper(seq) || seq;
  return {
    result: out,
    then: newMapper => conditionally(out, newMapper),
  };
}
