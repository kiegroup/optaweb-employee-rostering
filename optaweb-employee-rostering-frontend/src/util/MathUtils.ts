export function modulo(n: number, base: number) {
  const positiveBase = Math.abs(base);
  return ((n % positiveBase) + positiveBase) % positiveBase;
}
