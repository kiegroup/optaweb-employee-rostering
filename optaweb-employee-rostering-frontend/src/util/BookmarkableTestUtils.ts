
import * as router from 'react-router';
import { UrlProps } from './BookmarkableUtils';

export function getRouterProps<T extends UrlProps<any>>(pathname: string,
  props: Partial<T>): router.RouteComponentProps {
  const searchParams = new URLSearchParams();
  Object.keys(props).forEach((key) => {
    const value = props[key] as string | null | undefined;
    if (value) {
      searchParams.set(key, value);
    }
  });
  const location = {
    pathname,
    search: Object.keys(props).length > 0 ? `?${searchParams.toString()}` : '',
    hash: '',
    state: undefined,
  };
  const history = {
    location,
    push: jest.fn(),
    replace: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    block: jest.fn(),
    listen: jest.fn(),
    createHref: jest.fn(),
    action: 'PUSH' as any,
    length: 1,
  };

  jest
    .spyOn(router, 'useHistory')
    .mockImplementation(() => history);

  return {
    history,
    location,
    match: {
      isExact: true,
      path: pathname,
      url: `localhost:8080${pathname}`,
      params: {},
    },
  };
}
