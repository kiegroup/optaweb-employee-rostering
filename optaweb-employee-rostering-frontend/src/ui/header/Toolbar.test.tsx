import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { List } from 'immutable';
import { ToolbarComponent, Props } from './Toolbar';

describe('Toolbar Component', () => {
  beforeAll(() => {
    process.env.REACT_APP_BACKEND_URL = 'backend';
  });

  it('should render correctly with no tenants', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...noTenants} />);
    expect(toJson(toolbarComponent)).toMatchSnapshot();
  });

  it('should render correctly with a few tenants', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    expect(toJson(toolbarComponent)).toMatchSnapshot();
  });

  it('should render correctly when tenant select is open', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    (toolbarComponent.instance() as ToolbarComponent).setIsTenantSelectOpen(true);
    expect(toJson(toolbarComponent)).toMatchSnapshot();
  });

  it('should refresh tenant list on mount', async () => {
    shallow(<ToolbarComponent {...twoTenants} />);
    expect(twoTenants.refreshTenantList).toBeCalled();
  });

  it('should call change tenant on tenant change', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    (toolbarComponent.instance() as ToolbarComponent).setCurrentTenant(2);
    expect(twoTenants.changeTenant).toBeCalled();
    expect(twoTenants.changeTenant).toBeCalledWith({ routeProps: twoTenants, tenantId: 2 });
  });

  it('should redirect to admin page when you click on the gear', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    toolbarComponent.find('[aria-label="Settings"]').simulate('click');
    expect(twoTenants.history.push).toBeCalled();
    expect(twoTenants.history.push).toBeCalledWith('/admin');
  });
});

const noTenants: Props = {
  tenantList: List(),
  currentTenantId: 0,
  refreshTenantList: jest.fn(),
  changeTenant: jest.fn(),
  ...getRouterProps('/toolbar', {}),
};

const twoTenants: Props = {
  tenantList: List([
    {
      id: 1,
      version: 0,
      name: 'Tenant 1',
    },
    {
      id: 2,
      version: 0,
      name: 'Tenant 2',
    },
  ]),
  currentTenantId: 2,
  refreshTenantList: jest.fn(),
  changeTenant: jest.fn(),
  ...getRouterProps('/toolbar', {}),
};
