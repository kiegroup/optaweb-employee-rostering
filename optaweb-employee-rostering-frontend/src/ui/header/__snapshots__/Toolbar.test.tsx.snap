// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Component should render correctly when tenant select is open 1`] = `
<Toolbar>
  <ToolbarGroup>
    <ToolbarItem>
      <Dropdown
        dropdownItems={
          Array [
            <DropdownItem
              data-tenantid={1}
            >
              Tenant 1
            </DropdownItem>,
            <DropdownItem
              data-tenantid={2}
            >
              Tenant 2
            </DropdownItem>,
          ]
        }
        isOpen={true}
        isPlain={true}
        onSelect={[Function]}
        position="right"
        toggle={
          <DropdownToggle
            onToggle={[Function]}
          >
            Tenant 2
          </DropdownToggle>
        }
      />
    </ToolbarItem>
  </ToolbarGroup>
  <ToolbarGroup>
    <ToolbarItem>
      <Component
        aria-label="Notifications"
        variant="plain"
      >
        <BellIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </ToolbarItem>
    <ToolbarItem>
      <Component
        aria-label="Settings"
        data-cy="settings"
        onClick={[Function]}
        variant="plain"
      >
        <CogIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </ToolbarItem>
  </ToolbarGroup>
</Toolbar>
`;

exports[`Toolbar Component should render correctly with a few tenants 1`] = `
<Toolbar>
  <ToolbarGroup>
    <ToolbarItem>
      <Dropdown
        dropdownItems={
          Array [
            <DropdownItem
              data-tenantid={1}
            >
              Tenant 1
            </DropdownItem>,
            <DropdownItem
              data-tenantid={2}
            >
              Tenant 2
            </DropdownItem>,
          ]
        }
        isOpen={false}
        isPlain={true}
        onSelect={[Function]}
        position="right"
        toggle={
          <DropdownToggle
            onToggle={[Function]}
          >
            Tenant 2
          </DropdownToggle>
        }
      />
    </ToolbarItem>
  </ToolbarGroup>
  <ToolbarGroup>
    <ToolbarItem>
      <Component
        aria-label="Notifications"
        variant="plain"
      >
        <BellIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </ToolbarItem>
    <ToolbarItem>
      <Component
        aria-label="Settings"
        data-cy="settings"
        onClick={[Function]}
        variant="plain"
      >
        <CogIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </ToolbarItem>
  </ToolbarGroup>
</Toolbar>
`;

exports[`Toolbar Component should render correctly with no tenants 1`] = `
<Toolbar>
  <ToolbarGroup />
  <ToolbarGroup>
    <ToolbarItem>
      <Component
        aria-label="Notifications"
        variant="plain"
      >
        <BellIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </ToolbarItem>
    <ToolbarItem>
      <Component
        aria-label="Settings"
        data-cy="settings"
        onClick={[Function]}
        variant="plain"
      >
        <CogIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </ToolbarItem>
  </ToolbarGroup>
</Toolbar>
`;
